Handle type variable and polymorphic variant in the lexer.

hhugo opened the pull request:
Handles 'typ and \`variant in the lexer instead of the parser.

In particular, this PR disallows
- \`[SPACE]variant
- \`(* comment *)variant   
- \'[SPACE]typ   
- \'(* comment *)typ

hhugo committed:
Change lexer for quote ident

hhugo committed:
bootstrap parser

gasche commented on parsing/parser.mly:
```diff
@@ -2498,8 +2498,8 @@ with_type_binder:
-        QUOTE mkrhs(ident)                      { [$2] }
-      | typevar_list QUOTE mkrhs(ident)         { $3 :: $1 }
+        mkrhs(QUOTEIDENT)                      { [$1] }
+      | typevar_list mkrhs(QUOTEIDENT)         { $2 :: $1 }
```
Note that this changes the location attributed by `mkrhs` (probably in a good direction).

gasche commented on parsing/lexer.mll:
```diff
@@ -512,6 +520,14 @@ rule token = parse
+  | '`' anycase identchar *
+      { BACKQUOTEIDENT (get_quote_name lexbuf) }
+  | '`' anycase_latin1 identchar_latin1 *
+      { warn_latin1 lexbuf; BACKQUOTEIDENT (Lexing.lexeme lexbuf) }
```
Why are you dropping the backquote with `get_quote_name` in the `anycase` case but not `anycase_latin1`? (Same with QUOTEIDENT below). Note that you don't need to write `String.sub 1 ...` for this, you should be able to use

```
| '`' ((anycase identchar*) as ident)

hhugo replied:
This was a mistake due to multiple iterations. Fixed.

hhugo commented:
You've done some review on this already. Do you want some additional people to review as well?

gasche reviewed (commented):
Unless I missed something, I still believe that the `get_quote_name` function is superfluous; you should be able to use the `(anycase indentchar *) as ident` pattern in the lexing rule to capture the identifier. This is the way DOTOP is handled for example, and it gives lighter/simpler code.

If you want to factorize the lexical-error-reporting-code across the `anycase` and `anycase_latin1` branches, I would just have two functions `check_quoteident_name` and `check_backquoteident_name` that raise or return unit.


Otherwise I believe that the PR is fine. I agree with the change, but I'm a bit hesitant to approve of a lexical-syntax change (we rarely change it), so I would like to get the opinion of someone who cares about it, typically @damiendoligez.

hhugo commented:
> Unless I missed something, I still believe that the get_quote_name function is superfluous; you should be able to use the (anycase indentchar *) as ident pattern in the lexing rule to capture the identifier. This is the way DOTOP is handled for example, and it gives lighter/simpler code.

I kept the code similar to what's done for `LABEL` and `OPTLABEL` because the handling is very similar.

alainfrisch commented:
> In particular, this PR disallows

Is that the rationale for this PR or just a decent side-effect of it?

Would it now be possible to allow keywords as type variables or polymorphic variants?

hhugo commented:
The rational for this PR is was to make `' a'` (QUOTE SPACE CHAR QUOTE) not a valid token and not a valid type variable.  Such code was found here https://github.com/FStarLang/FStar/blob/v0.9.0/examples/low-level/Heap.ml#L45 while trying to reformat the source with ocamlformat.

The code responsible for pretty printing source code in the compiler also generate invalid source.
```ocaml
# type ' a' t = ' a';;
type 'a' t = 'a'
```

hhugo commented:
> Would it now be possible to allow keywords as type variables or polymorphic variants?

Yes, but it's not clear it is desirable.

pmetzger commented:
> Yes, but it's not clear it is desirable.

I'd say it is undesirable...

alainfrisch commented:
> The rational for this PR

Thanks for making this explicit.

The presence of code such as the one you mention should make us careful.

Is it still possible, after this PR, to use `'` in the name of a type variable?  If it has more than one identchar before the `'` (in which case it cannot be confused with a char literal)?

hhugo commented:
>Is it still possible, after this PR, to use ' in the name of a type variable? If it has more than one identchar before the ' (in which case it cannot be confused with a char literal)?

Yes, still possible.

> The presence of code such as the one you mention should make us careful.

I found only two occurrences in the entire opam-repository.
The other one is in the same directory https://github.com/FStarLang/FStar/blob/v0.9.0/examples/low-level/Lref.ml#L44.

gasche reviewed (dismissed):
Approved on my side. @alainfrisch, I'll let you either merge or voice more doubts.

alainfrisch commented:
>> Is it still possible, after this PR, to use ' in the name of a type variable? If it has more than one identchar before the ' (in which case it cannot be confused with a char literal)?
> Yes, still possible.

Ok, let's try to be explicit:

1. It is still ok to have ' in type variable names, but only if there are more than one letter before it:
````ocaml
# type 'a' t = unit;;
Error: Syntax error
# type 'aa' t = unit;;
type 'aa' t = unit
# type 'aa'' t = unit;;
type 'aa'' t = unit
````

So previously valid variable names (which could be written with whitespace before the initial ' and the name) are now disallowed, but this affects only single-letter+' .  Not dramatic, but this introduces some weird constraint on which AST can be pretty-printed to valid syntax.

2.  Another consequence is unrelated to type variables.  On trunk we can have a character literal followed immediately without whitespaces by another expression:
````ocaml
let f x y = (x,y) in f 'a'3;;
let f x y = (x,y) in f 'a''a';;
let f x y = (x,y) in f 'a'f;;
````
and after this PR all these become a syntax error.

I'm not fully convinced that the new situation is significantly better than the previous one, or at least not enough to justify the potential breakage.  So I won't take a decision on that, but if @gasche is confident enough, I won't fight against it.

Do I understand correctly that technically, the only current "bug" is that pprintast does not add a whitespace after the ' in some cases where this is necessary?

hhugo commented:
Thanks for pointing out 2.
This breaking change is somewhat similar to the time we disallowed integers to be immediately followed by `in` (without space). 
My feeling is that it's wrong to let the quote live potentially far away from the rest. 
```
`
(* I find this polymorphic variant very confusing *)
Polymorphic_variant
```

trefis commented:
Personally I think we're in better world if all the examples you give in point 2 end up being rejected. (And *I* definitely don't care about breaking such programs)

alainfrisch commented:
> My feeling is that it's wrong to let the quote live potentially far away from the rest.

Perhaps, but fixing that introduces other irregularities and could break existing code.  I'm questioning whether it's worth taking the risk.  But again, I'll let others decide.

(My opinion is that the real problem was allowing ' in identifier literals (but of course fixing that would break too much code now).)

hhugo committed:
update ocamllex lexer

hhugo commented:
I've now updated the ocamllex lexer accordingly.

hhugo commented:
I grepped all sources available in the opam repository with the following regexp
```
[^a-zA-Z']'[a-zA-Z_]'[a-zA-Z0-9']
```
- `.mli` files, nothing interesting, matches in comments.
- `.ml` files, few interesting matches, but all using `pa_ulex` and `mikmatch` (camlp4 syntax extensions)
- `.mll` files, lots of matches (e.g. `['a'-'z''A'-'Z''0'-'9']`), but skimming at them, no matches in productions.
- `.mly` files, no match

hhugo committed:
boostrap

hhugo committed:
Changelog

hhugo committed:
misc

hhugo commented:
@gasche, I've rebased on top of the lexer cleanup and updated the PR to use aliases in patterns.
@alainfrisch, based on my previous comment, do you still think the risk is high ?

garrigue commented:
Wait a minute.
I didn't answer yet on this PR, but I had marked for review to myself...

What this PR proposes is to change the specification of what is a type variable:
https://caml.inria.fr/pub/docs/manual-ocaml/types.html

Until now a type variable is defined not as a lexical entity but as a backquote followed by an identifier. Same thing for polymorphic variants.

While I agree that the parser is only an imperfect implementation of the reference manual, I don't think we're free to introduce incompatibilities arbitrarily, in particular when the goal is not to fix another incompatibility.

If you really think that these patterns as bad programming style, then we could have a warning.
While it would still be a bit dubious to warn in the only way to print a valid type variable name, one could argue that type variable names being only local, this is just warning about a bad choice of name.

garrigue reviewed (changes requested):
According to the reference manual, this should be a warning, not an error.

mshinwell commented:
If these patterns are bad style, why don't we fix the reference manual?

To my mind, all of the patterns in the original description of this pull request should obviously be disallowed, for clarity.

garrigue commented:
To clarify a bit things (as this may be an indirect cause for this PR), the reason `~label:` is lexicalized is not to proactively prohibit spaces between `label` and the symbols, but because not doing so would create an ambiguity on `(f ~label : t)` where `t` could otherwise be either an expression or a type. Limitations of parsing technology resulting in this not very clean solution.
By the way, in the non ambiguous case, you are allowed to write `f ~(*boo*)label` for an application with a reflexive label, following the specification.

alainfrisch commented:
> @alainfrisch, based on my previous comment, do you still think the risk is high ?

The risk is perhaps ok (although you found out a few problems in the wild, both for 1 and 2).  But I'm still not very fond of the irregularity (regarding which strings are valid type variable names in the AST).  This might be a cosmetics concern, but so is the one fixed by the PR.  Concretely, if one wants to do some quickcheck-style testing of Parsetree->syntax->Parsetree roundtrip, one would need to be careful about not generating `a'` as type variable names.

So I maintain my position: I'm not fully convinced, I will not be the one arguing in favor of merging, but I won't fight against it either.  **If** the PR is accepted, I'd still argue that the manual and `Ast_invariants` should be adapted.

garrigue commented:
@mshinwell The reference manual is also intended as a specification. If you change it, this should be argued first.
Honestly, I don't see the problem this PR is trying to solve.

gasche commented:
I withdrew my approval because @alainfrisch and @garrigue have voiced specific and reasonable doubts.

My understanding is that this PR is proposed with the intent of making [ocamlformat](https://github.com/ocaml-ppx/ocamlformat) a better tool. I think that ocamlformat is bound to become a very important tool in the OCaml community, so any PR about it starts with a big plus as far as I am personally concerned. On the other hand, the criticism on regularity of @alainfrisch and on the specification of @garrigue are also reasonable.

mshinwell commented:
I believe this PR is ultimately about `ocamlformat`, which while it may not be to everyone's taste, is a tool that appears to be gathering substantial support in the community.  I think we would do well to support it.

It seems to me that the point about the specification is splitting hairs somewhat.  If this is a good change to the language, then it seems like it should be changed, and the specification updated.  Is the concern about such a change to the specification breaking other people's work?  Which work would that be?

(Aside: It's unclear to me, in fact, why the formally-specified portions of the language grammar should not be replicated from the lexer/parser into the reference manual.  It's about the only part we _do_ have a formal specification for!)

In terms of which code examples are rejected consequentially to this change, it seems like all of the examples that have been presented here are of sufficiently poor style that they could reasonably be rejected.

I find @alainfrisch 's concern about "which AST can be pretty-printed to valid syntax" confusing.  Could an abstract type not be used to enforce the necessary constraint?

garrigue commented:
@mshinwell Could you then explain the relation with ocamlformat, and why it would require us to change this specific rule?
And no, it's not splitting hairs: if we don't even try to comply with the (partial) specification we have, we could just throw it away, or say that it is only documentation of the implementation and the implementation is the reference.
Also, while being able to test changes on opam is invaluable, I don't see it as a license to change the language just because the breakage is small. One reason is that not all software is opam, and that people might depend on the specification in more subtle ways than immediate breakage.

Finally, I think that PRs are about fixing things. Since there is no bug here, I don't see what it's fixing. If you want to fix the specification, then it would be better to first open a separate PR to discuss  that change (to the manual here), then you can fix the implementation to meet the specification.

alainfrisch commented:
> I find @alainfrisch 's concern about "which AST can be pretty-printed to valid syntax" confusing. Could an abstract type not be used to enforce the necessary constraint?

The problem is not about enforcing the constraint, but about adding an irregular, ad hoc constraint that needs to be specified and is, frankly, quite ugly (I'm talking about : "the quote character is allowed, but not in second position").  To be fair, we already have implicit and ad hoc constraints (the string cannot be an OCaml keyword).

I still don't see any description here of the problem with ocamlformat.  If it needs to print the type variable represented with "a'" in the Parsetree, just print it as `' a'`.  Yes, it's ugly, but the source code passed to ocamlformat already used that syntax.

Honestly, I feel we're all spending too much time on this.

mshinwell commented:
@garrigue Well, to my mind (and others my disagree) the fact that we accept any of the four examples at the top of this pull request (or indeed any of @alainfrisch 's number 2 examples about juxtaposition) is a bug.  I think the language should explicitly disallow constructs that on average are likely to be confusing, which I reckon those are.

@alainfrisch Why does it matter what the exact constraint is, though?  Presumably because users might fall foul of it.  So I would think about the chance that a user is going to hit the weird corner case, and then what happens if they do.  Isn't it very unlikely anyone would hit such a case, and if they did, we could just produce an error (which would be unlikely to be seen) explaining the constraint?

@hhugo Could you please provide a more detailed description as to the problem specifically as regards `ocamlformat`?

garrigue commented:
@mshinwell To me the problem is not that I agree or not on the above examples, but I disagree completely with your definition of bug. At least, when there is a specification, a bug is something that deviates from the specification. So technically this PR is introducing a bug where there was none. Now, the specification may be buggy, but this usually only qualifies contradictions in the specification itself. It may also contain misfeatures, which is probably what you are talking about here, and is indeed open to discussion. This is why I was suggesting opening a PR about the specification, not the implementation.

Also I do not think that it would be a good idea to turn the current implementation into a normative specification. While it may contain useful information for developers, it is also much less readable, and makes practical choices that we may want to revise eventually. Actually, at first I thought that this PR was just that, a change of the implementation without semantic consequences.

hhugo commented:
I think the discussion is taking the wrong direction here.

First, let me clarify, this PR has nothing to do with ocamlformat ! I don't know where the confusion came from.

I created this PR just because I found out about these confusing syntaxes and wanted to (maybe) improve the syntax in that area. This PR is about disallowing some confusing syntaxes (I'm sorry I did a very bad job a communicating that) and is not really about a specific implementation.

So like for most PRs, there are two different things to be discussed here. The general direction of the change and its implementation details.
It seems that most people here were happy with the general direction of the change and gave review on the implementation (@alainfrisch, I put your concern about irregularity in that bucket because it's a consequence of the current implementation).

@garrigue, I cannot quite tell where you stand. Some of your comments (`Honestly, I don't see the problem this PR is trying to solve`) seems to imply that you disagree with the general direction of the PR but your concerned seems to be more related to consequences of the current implementation.

> This might be a cosmetics concern, but so is the one fixed by the PR. Concretely, if one wants to do some quickcheck-style testing of Parsetree->syntax->Parsetree roundtrip, one would need to be careful about not generating `a'` as type variable names.

Note that pprintast is currently not printing type variable correctly.
   
> Finally, I think that PRs are about fixing things. Since there is no bug here, I don't see what it's fixing.

I really hope PRs are not only about fixing things :)

garrigue commented:
OK, you may be mixing my comments with Alain's.
I don't think I've said anything about the implementation.
The "ambiguity" in my comments is that I did not intend to attack your goal, but was just making the point that this PR as it is stated was actually introducing a bug (i.e. a deviation from the specification). While a PR may also be about improvements to the code base (and your cleanup of lexer.mll was welcome), the improvement should be demonstrated, and the PR itself should not contain bugs :-)

In the end, I think that my position on the specification itself would be the same: the fact one may write poorly readable code does not seem justification enough to change an otherwise regular specification. Combinations of symbols and letters are not lexicalized in the specification, and this is far from being the only cases. And even if we decide that lexicalizing is better, Alain's arguments would have to be considered.

alainfrisch commented:
It's a small change, but it affects the specification and possibly external tools, fixings some problems but introducing some others.  I think we should rather close it and fix pprintast (and ocamlformat) to produce valid code considering the current specification.

Or is any maintainer willing to champion this PR?

mshinwell commented:
@alainfrisch I think we all may have got the wrong end of the stick about how this is or is not related to `ocamlformat`, but as far as the rest of this PR goes, I stand by what I said in <https://github.com/ocaml/ocaml/pull/2034#issuecomment-424387099>: some of these allowed syntaxes seem sufficiently bad to be deemed as a bug (whether in the specification, implementation or both).

alainfrisch commented:
As Jacques mentioned:

> Combinations of symbols and letters are not lexicalized in the specification, and this is far from being the only cases.

Typically, one can write `f  ~ (* wait for it *) x`.  Is that bad style?  Yes, of course.  Should we change the spec to reject it: I'm not sure.

alainfrisch commented:
The real question is whether the current spec leads to hard-to-understand error messages or surprising behavior for users.  I don't expect users to use those forms in practice, so I don't think this is the case.

damiendoligez commented:
I'd like to get a warning on type variables with a quote as their second character (and indeed another warning on any quote inside an identifier). If you're ambitious, a warning on whitespace (or comments) between the quote (or backquote) and the ident.

But of course that would be a completely different implementation.

In its current state, this PR has:
- the (small) risk of breaking some code we don't know of
- the (small) inconvenience of introducing a strange constraint on type variable names
- the (small) benefit of enforcing a bit of good taste in code

As it stands, I don't think the benefits outweigh the costs.

gasche commented:
We also merged #2130 in the meantime, which fixes the problem of the problematic type variables being reprinted incorrectly. I think that there is a consensus here to move on and close this issue. (Apologies, @hhugo, for the work you put in that PR.)

hhugo commented:
I'm fine with the resolution. @gpetiot, this should probably be fixed in ocamlformat
